symengine/basic-inl.h

```diff
@@ -40,6 +40,53 @@ inline bool is_a_sub(const Basic &b)
     return dynamic_cast<const T *>(&b) != nullptr;
 }
 
+template <typename To, typename From>
+inline To implicit_cast(const From &f){
+    return f;
+}
+
+// debug safe static_cast.
+// use like this: down_cast<T*>(foo).
+// only accept pointers.
+template <typename To, typename From>
+inline To down_cast(From* f){
+    // Ensures that To is a sub-type of From *.  This test is here only
+    // for compile-time type checking, and has no overhead in an
+    // optimized build at run-time, as it will be optimized away
+    // completely.
+    if (false) {
+        implicit_cast<From*, To>(0);
+    }
+
+#if defined(TEUCHOS_DEBUG) 
+    assert(f == NULL || dynamic_cast<To>(f) != NULL);  // debug mode only!
```

Comment:
Use `SYMENGINE_ASSERT`, because the regular `assert` is enabled/disabled with `NDEBUG` I think.